dubbo-remoting/dubbo-remoting-zookeeper/src/main/java/org/apache/dubbo/remoting/zookeeper/support/AbstractZookeeperTransporter.java

```diff
@@ -0,0 +1,141 @@
+    private static final Logger logger = LoggerFactory.getLogger(ZookeeperTransporter.class);
+    private final Map<String, ZookeeperClient> zookeeperClientMap = new ConcurrentHashMap<>();
+
+    /**
+     * share connnect for registry, metadata, etc..
+     *
+     * @param url
+     * @return
+     */
+    public ZookeeperClient connect(URL url) {
+        ZookeeperClient clientData;
+        List<String> addressList = getURLBackupAddress(url);
+        // The field define the zookeeper server , including protocol, host, port, username, password
+        if ((clientData = fetchAndUpdateZookeeperClientCache(url, addressList)) != null) {
+            logger.info("Get result from map for the first time when invoking zookeeperTransporter.connnect");
+            return clientData;
+        }
+        ZookeeperClient zookeeperClient = null;
+        // avoid creating too many connections， so add lock
+        synchronized (zookeeperClientMap) {
+            if ((clientData = fetchAndUpdateZookeeperClientCache(url, addressList)) != null) {
+                logger.info("Get result from map for the second time when invoking zookeeperTransporter.connnect");
+                return clientData;
+            }
+
+            zookeeperClient = createZookeeperClient(createServerURL(url));
+            logger.info("Get result by creating new connection when invoking zookeeperTransporter.connnect. ");
+            writeToClientMap(addressList, zookeeperClient);
+        }
+        return zookeeperClient;
+    }
+
+    /**
+     * @param url the url that will create zookeeper connection .
+     *            The url in AbstractZookeeperTransporter#connect parameter is rewritten by this one.
+     *            such as: zookeeper://127.0.0.1:2181/org.apache.dubbo.remoting.zookeeper.ZookeeperTransporter
+     * @return
+     */
+    protected abstract ZookeeperClient createZookeeperClient(URL url);
+
+    ZookeeperClient fetchAndUpdateZookeeperClientCache(URL url, List<String> addressList) {
+
+        ZookeeperClient zookeeperClientData = null;
+        for (String address : addressList) {
+            if ((zookeeperClientData = zookeeperClientMap.get(address)) != null) {
+                break;
+            }
+        }
+        if (zookeeperClientData != null) {
+            writeToClientMap(addressList, zookeeperClientData);
+        }
+        return zookeeperClientData;
+    }
+
+    List<String> getURLBackupAddress(URL url) {
+        List<String> addressList = new ArrayList<String>();
+        addressList.add(url.getAddress());
+
+        String[] backups = url.getParameter(Constants.BACKUP_KEY, new String[0]);
+        addressList.addAll(Arrays.asList(backups));
+        return addressList;
+    }
+
+    /**
+     * write address-ZookeeperClient relationship to Map
+     *
+     * @param addressList
+     * @param ZookeeperClient
+     */
+    void writeToClientMap(List<String> addressList, ZookeeperClient ZookeeperClient) {
```

Comment:
we generally use camel case as variable naming convention. Should we user **zookeeperClient** for methor parameter?

Reply:
method definition can re written as 

```
private void  writeToClientMap(List<String> addressList, ZookeeperClient zookeeperClient){
  addressList.forEach(address->zookeeperClientMap.put(address, zookeeperClient));
}
```

Reply:
Modified